src/ring_theory/noetherian.lean

```diff
@@ -508,3 +511,242 @@ nat.rec_on n
+  replace h_PM : P M, from h_PM,
+  have h_not_prM : ¬ M.is_prime,
+  { by_contradiction hM,
+    specialize h_PM {⟨M, hM⟩},
+    rw [multiset.singleton_eq_singleton, multiset.map_cons,
+      multiset.map_zero, multiset.prod_singleton] at h_PM,
+    rw [subtype.val_eq_coe, subtype.coe_mk, le_iff_lt_or_eq,
+      not_or_distrib, eq_self_iff_true, not_true] at h_PM,
+    exact h_PM.right },
+  have h_not_topM : M ≠ ⊤,
+  { suffices h_top : ¬ P ⊤,
+    { apply mt (congr_arg P), tauto },
+    { rw not_forall,
+      obtain ⟨Ξ, hΞ⟩ : ∃ (Ξ : ideal R), Ξ.is_maximal,
+      apply_rules @ideal.exists_maximal _ _ _,
+      let pΞ : (prime_spectrum R) := ⟨Ξ, ideal.is_maximal.is_prime hΞ⟩,
+      use ({ pΞ } : multiset (prime_spectrum R)),
+      rw [not_not, multiset.singleton_eq_singleton, multiset.map_cons,
+        multiset.map_zero, multiset.prod_singleton],
+      simp only [le_top] } },
+  obtain ⟨x, y, hx, hy, h_xy⟩ : ∃ (x y : R), x ∉ M ∧ y ∉ M ∧ x * y ∈ M,
+  { rw [ideal.is_prime, not_and] at h_not_prM,
+    specialize h_not_prM h_not_topM,
+    rw not_forall at h_not_prM,
+    cases h_not_prM with x hx,
+    rw not_forall at hx,
+    cases hx with y hxy,
+    use [x, y], finish },
+  let Jx := M + span R {x},
+  let Jy := M + span R {y},
+  have this : Jx ∉ Ω ∧ Jy ∉ Ω,
+  { have incl : ∀ (z : R), M ≤ M + span R {z},
+    { intro z, from le_sup_left },
+    have belong : ∀ (z : R), z ∈ M + span R {z},
+    { intro z,
+      apply submodule.mem_sup_right,
+      apply submodule.mem_span_singleton_self z },
+    split, all_goals {by_contradiction},
+    { have abs_x : Jx = M, from h_maxM Jx h (incl x),
+      rw ← abs_x at hx, tauto },
+    { have abs_y : Jy = M, from h_maxM Jy h (incl y),
+      rw ← abs_y at hy, tauto } },
+  obtain ⟨Wx, h_Wx⟩ : ∃ (Wx : multiset (prime_spectrum R)), multiset.prod (Wx.map subtype.val) ≤ Jx,
+  { have Jx_P : ¬ P Jx, simp [*, mem_set_of_eq, not_forall] at *,
+    dsimp [P] at Jx_P, from not_forall_not.mp Jx_P },
+  obtain ⟨Wy, h_Wy⟩ : ∃ (Wy : multiset (prime_spectrum R)), multiset.prod (Wy.map subtype.val) ≤ Jy,
+  { have Jy_P : ¬ P Jy, simp [*, mem_set_of_eq, not_forall] at *,
+    dsimp [P] at Jy_P, from not_forall_not.mp Jy_P },
+  let W := Wx + Wy,
+  have h_W : multiset.prod (W.map subtype.val) ≤ M,
+  { suffices h_JJ : multiset.prod (W.map subtype.val) ≤ Jx * Jy,
+    { suffices h_JM : Jx * Jy ≤ M, from le_trans h_JJ h_JM,
+      { dsimp only *,
+        rw add_mul,
+        have this : M * (M + span R {y}) ≤ M, from ideal.mul_le_right,
+        apply sup_le this,
+        rw mul_add,
+        replace this : span R {x} * M ≤ M, from ideal.mul_le_left,
+        apply sup_le this,
+        rw [span_mul_span, singleton_mul_singleton, span_singleton_le_iff_mem],
+        exact h_xy } },
+    { rw [multiset.map_add, multiset.prod_add],
+      apply submodule.mul_le_mul h_Wx h_Wy }, },
+  have h_absM : ¬ P M,
+  { dsimp only [P], rw not_forall_not,
+    use W, exact h_W },
+  apply absurd h_PM h_absM,
+end
```

Comment:
We can split off two lemmas to shorten the proof. The first one can go in `ring_theory.ideal.basic`:
```lean
lemma not_is_prime_iff {I : ideal R} : ¬ I.is_prime ↔ I = ⊤ ∨ ∃ (x ∉ I) (y ∉ I), x * y ∈ I :=
begin
  simp_rw [ideal.is_prime, not_and_distrib, ne.def, not_not, not_forall, not_or_distrib],
  exact or_congr iff.rfl
    ⟨λ ⟨x, y, hxy, hx, hy⟩, ⟨x, hx, y, hy, hxy⟩, λ ⟨x, hx, y, hy, hxy⟩, ⟨x, y, hxy, hx, hy⟩⟩
end
```
The second one after `set_has_maximal_iff_noetherian`:
```lean
/-- If `∀ I > J, P I` implies `P J`, then `P` holds for all submodules. -/
lemma is_noetherian.induction {R M} [ring R] [add_comm_group M] [module R M] [is_noetherian R M]
  {P : submodule R M → Sort*} (hgt : ∀ I, (∀ J > I, P J) → P I)
  (I : submodule R M) : P I :=
well_founded.recursion (well_founded_submodule_gt R M) I hgt
```

With those, I managed to golf this down to:
```suggestion
begin
  refine is_noetherian.induction (λ (M : ideal R) hgt, _) I,
  by_cases h_prM : M.is_prime,
  { use {⟨M, h_prM⟩},
    rw [multiset.map_singleton, multiset.singleton_eq_singleton, multiset.prod_singleton,
        subtype.coe_mk],
    exact le_rfl },
  by_cases htop : M = ⊤,
  { rw htop,
    exact ⟨0, le_top⟩ },
  have lt_add : ∀ z ∉ M, M < M + span R {z},
  { intros z hz,
    refine lt_of_le_of_ne le_sup_left (λ m_eq, hz _),
    rw m_eq,
    exact mem_sup_right (mem_span_singleton_self z) },
  obtain ⟨x, hx, y, hy, hxy⟩ := (not_is_prime_iff.mp h_prM).resolve_left htop,
  let Jx := M + span R {x},
  let Jy := M + span R {y},
  obtain ⟨Wx, h_Wx⟩ := hgt (M + span R {x}) (lt_add _ hx),
  obtain ⟨Wy, h_Wy⟩ := hgt (M + span R {y}) (lt_add _ hy),
  use Wx + Wy,
  let W := Wx + Wy,
  rw [multiset.map_add, multiset.prod_add],
  apply le_trans (submodule.mul_le_mul h_Wx h_Wy),
  rw add_mul,
  apply sup_le (show M * (M + span R {y}) ≤ M, from ideal.mul_le_right),
  rw mul_add,
  apply sup_le (show span R {x} * M ≤ M, from ideal.mul_le_left),
  rwa [span_mul_span, singleton_mul_singleton, span_singleton_le_iff_mem],
end
```